### PR TITLE
Allow custom configuration interpolation logic

### PR DESCRIPTION
**Pull request by wendigo**

### CODE REVIEW

**Comment by wendigo** [comment]
cc @martint @electrum

File: configuration/src/main/java/io/airlift/configuration/ConfigurationInterpolator.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                String interpolatorName = matcher.group("name");
+                String keyName = matcher.group("key");
+                ValueInterpolator valueInterpolator = valueProviders.get(interpolatorName);
+                if (valueInterpolator == null) {
+                    onError.accept(propertyKey, "Configuration property '%s' references value interpolator '%s' that does not exist".formatted(propertyKey, interpolatorName));
+                    return;
+                }
+
+                Optional<String> replacedValue = valueInterpolator.interpolate(keyName);
+                if (replacedValue.orElse(null) == null) {
```
Comment (nineinchnick) [inline configuration/src/main/java/io/airlift/configuration/ConfigurationInterpolator.java:57]: Would `isEmpty()` be more idiomatic? Should we assert that `replacedValue` is not null? That would signal some other issues with the interpolator, right?
```suggestion
                if (replacedValue.isEmpty()) {
```

File: configuration/src/main/java/io/airlift/configuration/ConfigurationInterpolator.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            }
+            matcher.appendTail(replacedPropertyValue);
+            replaced.put(propertyKey, replacedPropertyValue.toString());
+        });
+        return replaced;
+    }
+
+    public interface ValueInterpolator
+    {
+        Optional<String> interpolate(String key);
```
Comment (nineinchnick) [inline configuration/src/main/java/io/airlift/configuration/ConfigurationInterpolator.java:71]: If this interface would also have a `name()` method, we could avoid passing `valueProviders` as a map and maybe use a set. I was wondering if the `ENV` key should be defined as a const, but then figured out that if we always associate it with `EnvValueInterpolator` it should be embedded in it.

**Comment by electrum** [comment]
What is the use case for this?

**Comment by wendigo** [comment]
@electrum interpolating secrets from other sources than env (i.e. AWS secretmanager or Hashicorp vault)

**Pull request closed without merging**